src/chkstat.cpp

```diff
@@ -21,1365 +21,1246 @@
             {
-                BAD_LINE();
-                continue;
+                // relative link: continue relative to the parent directory
+
+                // we encountered a link directly below /, simply continue
+                // from the root again
+                if (parentfd.invalid())
+                {
+                    pathfd.close();
+                }
+                else
+                {
+                    pathfd.set( dup(parentfd.get()) );
+                }
             }
-            last = add_permlist(part[0], part[1], part[2], mode);
+
+            path_rest = link + path_rest;
+        }
+        else if (ctx.status.isDirectory())
+        {
+            // parentfd is only needed to find the parent of a symlink.
+            // We can't encounter links when resolving '.' or '..' so those don't need any special handling.
+            parentfd.set( dup(pathfd.get()) );
         }
-        fclose(fp);
     }
 
-    euid = geteuid();
-    for (e = permlist; e; e = e->next)
+    // world-writable file: error out due to unknown file integrity
+    if (ctx.status.isRegular() && ctx.status.isWorldWritable())
     {
-        if (use_checklist && !in_checklist(e->file+rootl))
-            continue;
+        std::cerr << ctx.subpath << ": file has insecure permissions (world-writable)" << std::endl;
+        return false;
+    }
 
-        pwd = strcmp(e->owner, "unknown") ? getpwnam(e->owner) : NULL;
-        grp = strcmp(e->group, "unknown") ? getgrnam(e->group) : NULL;
-        uid = pwd ? pwd->pw_uid : 0;
-        gid = grp ? grp->gr_gid : 0;
+    ctx.fd.steal(pathfd);
 
-        bool traversed_insecure;
-        if (fd >= 0)
-        {
-            // close fd from previous loop iteration
-            close(fd);
-            fd = -1;
-        }
+    return true;
+}
+
+std::string Chkstat::getPathFromProc(const FileDesc &fd) const
+{
+    std::string linkpath;
+    auto procpath = std::string("/proc/self/fd/") + std::to_string(fd.get());
+
+    linkpath.resize(PATH_MAX);
 
-        fd = safe_open(e->file, &stb, uid, &traversed_insecure);
-        if (fd < 0)
+    ssize_t l = readlink(procpath.c_str(), &linkpath[0], linkpath.size() - 1);
+    if (l > 0)
+    {
+        linkpath.resize( std::min(static_cast<size_t>(l), linkpath.size() - 1) );
+    }
+    else
+    {
+        linkpath = "ancestor";
+    }
+
+    return linkpath;
+}
+
+int Chkstat::processEntries()
+{
+    EntryContext ctx;
+    size_t errors = 0;
+
+    if (m_apply_changes.isSet() && !checkHaveProc())
+    {
+        std::cerr << "ERROR: /proc is not available - unable to fix policy violations. Will continue in warn-only mode." << std::endl;
+        errors++;
+        m_apply_changes.setValue(false);
+    }
+
+    for (auto &pair: m_profile_entries)
+    {
+        // this needs to be non-const currently, because further below the
+        // getCapabilities logic is modifying entry properties on the fly.
+        auto &entry = pair.second;
```

Comment:
`pair` can be destructured directly in the loop header since C++17: https://en.cppreference.com/w/cpp/language/structured_binding#Example

Reply:
Oh this is cool. I've always missed in in C++11.